include/pybind11/chrono.h

```diff
@@ -0,0 +1,177 @@
+                + seconds(delta->seconds)
+                + microseconds(delta->microseconds));
+            return true;
+        }
+        // If they have passed us a float we can assume it is seconds and convert
+        else if (PyFloat_Check(src.ptr())) {
+            double val = PyFloat_AsDouble(src.ptr());
+            // Multiply by the reciprocal of the ratio and round
+            value = type(std::lround(val * type::period::den / type::period::num));
+            return true;
+        }
+        else return false;
+    }
+
+    static handle cast(const std::chrono::duration<Rep, Period> &src, return_value_policy /* policy */, handle /* parent */) {
+        using namespace std::chrono;
+        if (!PyDateTimeAPI) { PyDateTime_IMPORT; }
+
+        // Declare these special duration types so the conversions happen with the correct primitive types (int)
+        using dd_t = duration<int, std::ratio<86400>>;
+        using ss_t = duration<int, std::ratio<1>>;
+        using us_t = duration<int, std::micro>;
+
+        return PyDelta_FromDSU(
+              duration_cast<dd_t>(src).count()
+            , duration_cast<ss_t>(src % days(1)).count()
+            , duration_cast<us_t>(src % seconds(1)).count());
+    }
+    PYBIND11_TYPE_CASTER(type, _("datetime.timedelta"));
+};
+
+// This is for casting times on the system clock into datetime.datetime instances
+template <typename Duration> class type_caster<std::chrono::time_point<std::chrono::system_clock, Duration>> {
+public:
+    typedef std::chrono::time_point<std::chrono::system_clock, Duration> type;
+    bool load(handle src, bool) {
+        using namespace std::chrono;
+
+        // Lazy initialise the PyDateTime import
+        if (!PyDateTimeAPI) { PyDateTime_IMPORT; }
+
+        if (!src) return false;
+        if (PyDateTime_Check(src.ptr())) {
+            std::tm cal;
+            cal.tm_sec   = PyDateTime_DATE_GET_SECOND(src.ptr());
+            cal.tm_min   = PyDateTime_DATE_GET_MINUTE(src.ptr());
+            cal.tm_hour  = PyDateTime_DATE_GET_HOUR(src.ptr());
+            cal.tm_mday  = PyDateTime_GET_DAY(src.ptr());
+            cal.tm_mon   = PyDateTime_GET_MONTH(src.ptr()) - 1;
+            cal.tm_year  = PyDateTime_GET_YEAR(src.ptr()) - 1900;
+            cal.tm_isdst = -1;
+
+            value = system_clock::from_time_t(mktime(&cal)) + microseconds(PyDateTime_DATE_GET_MICROSECOND(src.ptr()));
+            return true;
+        }
+        else return false;
+    }
+
+    static handle cast(const std::chrono::time_point<std::chrono::system_clock, Duration> &src, return_value_policy /* policy */, handle /* parent */) {
+        using namespace std::chrono;
+
+        // Lazy initialise the PyDateTime import
+        if (!PyDateTimeAPI) { PyDateTime_IMPORT; }
+
+        time_t tt = system_clock::to_time_t(src);
+        // this function uses static memory so it's best to copy it out asap just in case
+        tm *ltime = localtime(&tt);
```

Comment:
I made it a little simpler, but the comment about copying out still stands as localtime could be used outside of python without the protection of the GIL. localtime has a static memory block inside that it stores the value in and when it's called again it uses the same memory. There is a posix standard function localtime_r that doesn't do this, and on windows localtime uses thread local memory so is thread safe. If you can work out a way to switch between them that would help me sleep better at night, SFINAE doesn't work on c functions :(